Feat: Sidebar 컴포넌트 생성

yejineee opened the pull request:
## 구현 내용
-  마일스톤, 라벨, assignee의 내용을 담은 사이드바 구현

## 알림
- 선택된 마일스톤, 라벨, assignee는 어떠한 이름으로 보낼지 아직 결정이 안되어있는데, 
결정된다면 알려주세요!
  - 마일스톤은 Number or undefined, 라벨과 assignee는 Set([]) 구조로 보내주시면 됩니다!
- 사이드바 모델로 handler를 보내실 때, 다음과 같은 구조로 보내주시면 됩니다!
```jsx
handlers{
  updateMilestone,
  updateLabel,
  updateAssignee,
}
```
## 사이드바 화면
<img width="315" alt="스크린샷 2020-11-10 오후 7 53 33" src="https://user-images.githubusercontent.com/43772082/98667867-5d278300-2392-11eb-8f4b-f67dbcf07c8c.png">

**수정**
- 리뷰사항을 반영해서 css부분에 몇 가지 수정을 하였습니다. 
1. border는 삭제하고 사이드바 사이에 &+& 셀렉터로 border-top에만 선을 그었습니다.
2. 드롭다운 아이템은 width 100%로 두고, height는 삭제하고, padding으로 height를 설정하였습니다.
3. cursor는 hover 밖에 두었습니다.
- sidebar
<img width="1407" alt="스크린샷 2020-11-11 오전 7 09 45" src="https://user-images.githubusercontent.com/43772082/98739664-1b77f600-23ed-11eb-94ed-ff780fbd86e9.png">

- dropdown 
<img width="314" alt="스크린샷 2020-11-11 오전 10 57 22" src="https://user-images.githubusercontent.com/43772082/98755883-c77d0980-240c-11eb-98fb-6047b8d37631.png">

yejineee committed:
Feat: 마일스톤 진행률 컴포넌트 추가 (#125)

* Feat: 마일스톤의 진행률을 보여주는 유틸함수 추가

마일스톤에서 시작 시간부터 끝 시간 사이에서 현재 시간은 몇 %의 진행률인지를
보여주는 유틸리티 함수를 추가하였다.
현재시간이 끝시간보다 크다면, 100을 반환한다.
시작시간이 끝 시간보다 크다면, 에러를 반환한다.
위 두 경우를 제외하고는 진행률을 % 단위로 반환한다.

* Test: 마일스톤 진행률 테스트 추가

마일스톤 진행률을 보여주는 유틸리티 함수의 테스트를 추가하였다.

* Feat: 마일스톤 진행률 컴포넌트 추가

시작시간과 끝 시간이 주어졌을 때, 현재 시간상 진행률을 보여주는
컴포넌트를 추가하였다.

Co-authored-by: yejineee <21700431@handong.edu>

rolled-potatoes committed:
마크다운 에디터 상태 분리 (#124)

* Refactor : 마크다운 에디터 상태 분리

LeeSuKyeong committed:
Feat: sidebar 레이아웃 구현 (#126)

사이드바 구현에 필요한 컴포넌트 구현.

Commit:
Fix: milestone dummy data field 이름 수정

실제 api 구현 시 endDate로 데이터가 오기 때문에,
더미데이터도 이와 맞춰 end_date로 되어있는 것을 endDate로 수정한다.

Commit:
Fix: milestone progress를 실제 데이터를 props로 받는 것으로 수정

milestone을 가져오는 api를 연결하여 실제 데이터를 가져왔으므로,
더미데이타를 사용했던 것을 props로 가져온 데이터를 사용하는 것으로 수정한다.

Commit:
Chore: Rename TextDropdownItem -> FilterDropdownItem

TextDropDownItem을 필터와 마일스톤 두 곳에서 범용적으로
사용하는데 문제가 있어, 필터 드롭다운으로 이름을 변경한다.

Commit:
Style: dropdown을 button으로 수정

dropdown 아이템을 클릭했을 시, 클릭 이벤트가 발생해야 하므로
DropDownItemContainer를 button으로 수정한다.
버튼으로 변경된 후, 스타일을 조정하였다.

Commit:
Fix: 드롭다운에 click handler, dataset-id, 조건부 아이콘 렌더링

각 드롭다운마다 고유한 id가 무엇인지 설정하여, 클릭 이벤트 발생 시
이벤트 핸들러에서 id를 알 수 있도록 설정하였다.
또한, 드롭다운마다 클릭 이벤트 핸들러를 설정하였다.
드롭다운이 선택되었을 경우에만, 체크 아이콘이 보이도록 설정하였다.

Commit:
Feat: 프론트에서 서버로 데이터 가져오는 api 생성

마일스톤, 라벨, 유저 리스트를 프론트에서 서버로 데이터를 가져오는
api를 추가하였다. 이는 사이드바에서 사용이 된다.

Commit:
Feat: 마일스톤, 유저, 라벨이 선택된 여부에 따라 렌더링

마일스톤과 유저, 라벨은 선택되지 않았을 경우
no milestones와 같은 메시지를 띄운다.
만약 선택된 것이 있는 경우, 선택된 것을 렌더링해주도록 한다.

Commit:
Feat: 사이드바 컴포넌트 구성

assignee, labels, milestones 세 가지의 요소를 합쳐
사이드바를 생성하는 사이드바 컴포넌트를 구현하였다.

Commit:
Feat: 사이드바 모델 생성

사이드바 모델에서는 labels, assignees, milestones를
서버로부터 가져와서 사이드바 뷰에 데이터를 props로 넘겨준다.

Commit:
Feat: 사이드바에서 fetch 성공시 리듀서, action-type 추가

사이드바에서는 label, assignee, milestone을 갖고 있다.
각 경우마다 fetch에 성공했을 시, state에 추가하는 리듀서를 구현하였다.

changheedev commented on front/src/libs/utils.test.js:
```diff
@@ -14,3 +15,27 @@ test('리스트에서 조건에 맞는 엘리먼트의 갯수를 가져온다. '
     1,
   );
 });
+
+describe('현재시간보다 지났는지 확인할 때,', () => {
+  test('현재시간이 기준 시간보다 지났으면 true를 반환한다.', () => {
+    const date = '2020-11-10 13:27:11';
+    expect(utils.isTimePassedFromNow(date)).toBeTruthy();
+  });
+  test('현재시간이 기준 시간을 지나지 않았으면 false를 반환한다.', () => {
+    const date = '2022-11-10 13:28:11';
+    expect(utils.isTimePassedFromNow(date)).toBeFalsy();
+  });
+});
+describe('시작 시간부터 끝 시간을 확인할 때,', () => {
```
테스트 굳굳 👍

changheedev reviewed (commented):
마크다운 에디터 변경된 내용은 이전 PR에서 이미 merge가 된것 같은데 여기서도 변경점으로 잡히고 있네요.

changheedev commented on front/src/components/commons/MilestoneProgress.js:
```diff
@@ -0,0 +1,33 @@
+`;
+
+const MilestoneProgress = ({ createdAt, endDate }) => {
+  const progressRate = utils.getProgressRate(createdAt, endDate);
+  console.log(progressRate);
+  return (
```
디버깅용 console.log 가 !!

yejineee replied:
바~로 삭제했습니다!

changheedev commented on front/src/components/commons/dropdown/DropdownItemContainer.js:
```diff
@@ -1,14 +1,17 @@
   padding: 0.2rem 1rem;
-  height: 1.5rem;
+  width: 18rem;
+  height: 3rem;
```
이 부분은 드롭다운에 들어가는 리스트에서 각 아이템의 width와 height이 되는건가요??
- width는 방법의 차이겠지만 아이템 리스트를 감싸는 상위 컨테이너에서 설정해주고 여기서는 100%로 설정하는 방법도 있습니다. 저는 이 방식을 좀 더 선호하는 편입니다.
- height 은 따로 설정 하지 않은 상태에서 padding을 추가하게 되면 적당한 크기가 될 것 같아요.

changheedev commented on front/src/components/commons/dropdown/DropdownItemContainer.js:
```diff
@@ -1,14 +1,17 @@
-const DropdownItemContainer = styled.div`
+const DropdownItemContainer = styled.button`
   display: flex;
+  border: none;
   border-top: 1px solid ${color.lightGray};
```
![스크린샷 2020-11-10 오후 9 12 13](https://user-images.githubusercontent.com/17294694/98672645-73850d00-2399-11eb-8889-701e9487d19b.png)

위에 스크린샷을 보면 제일 첫 아이템의 border-top은 사실 그 위에 있는 요소의 border-bottom으로 처리될 것 같아요.
그래서 아이템끼리의 사이에만 border를 설정해주면 되지 않을까 싶습니다. 
```
& + & {
  border-top: 1px solid ${color.lightGray};
}
```

yejineee replied:
&+& 참고해서 수정했습니다 ㅎ-ㅎ!!! 감사합니다

changheedev commented on front/src/components/commons/sidebar/SidebarContainer.js:
```diff
@@ -0,0 +1,142 @@
+  flex-direction: column;
+  align-items: stretch;
+  width: 100%;
+  padding: 0.5rem 0;
+  border: 1px solid green;
+`;
```
![스크린샷 2020-11-10 오후 9 17 26](https://user-images.githubusercontent.com/17294694/98673076-281f2e80-239a-11eb-9460-c79803e660c7.png)

이 부분은 임시로 border를 주신 걸 수도 있겠지만 border가 겹쳐서 두꺼워지는 것을 볼 수 있는데 컨테이너 사이의 border는 `+` 선택자로 처리하고 테두리 border는 상위 컨테이너에 주면 더 깔끔할 것 같습니다.

yejineee replied:
위의 css 관련 리뷰사항 다 반영했습니다! + 셀렉터는 잘 사용 안했었는데, 덕분에 유용하게 사용하는 것 같습니다! 감사합니다!!

yejineee replied:
임시로 둔 border라서 삭제하고, &+& border-top에 Gray 컬러로 두었습니다! 위의 PR message의 사진에서 확인해주세용 ㅎㅎ 감사합니다!

rolled-potatoes commented on front/src/components/commons/dropdown/DropdownItemContainer.js:
```diff
@@ -1,14 +1,17 @@
+  width: 18rem;
+  height: 3rem;
+  background: transparent;
   &:hover {
     background: ${color.lightBlue};
+    cursor: pointer;
   }
```
`cursor : pointer` 는 밖에 있어도 될꺼 같아요 !

rolled-potatoes commented on front/src/components/commons/sidebar/SelectedAssignee.js:
```diff
@@ -0,0 +1,41 @@
+    <>
+      {selectedAssignees.size ? (
+        <div>
+          {[...selectedAssignees].map((selectedAssigneeId) => {
+            const assignedUser = assignees.find(
+              (assignee) => +assignee.id === +selectedAssigneeId,
+            );
+            return (
+              assignedUser && (
+                <div>
+                  <UserProfile src={assignedUser.profileLink} size="sm" />
+                  <span>{assignedUser.name}</span>
+                </div>
+              )
+            );
+          })}
+        </div>
```
렌더링 되는 부분에서 좀 복잡해진감이 있는거 같아요.
상위에 간단한 컴포넌트로 만드는건 어떨 까요?

yejineee replied:
분리시켰습니다!

yejineee replied:
한 단계 더 분리하였어용!

rolled-potatoes commented on front/src/models/sidebar/index.js:
```diff
@@ -0,0 +1,75 @@
+const Sidebar = ({ handlers }) => {
+  const { updateMilestone, updateLabel, updateAssignee } = handlers;
+
+  const getLabelsApi = () => labelApi.getLabels();
+  const getMilestonesApi = () => milestoneApi.getMilestones();
+  const getAssgineesApi = () => usersApi.getUsers();
+
```
이거를 따로 변수에 함수로 만든 이유가 있는가요 ?

yejineee replied:
인자값을 받는 경우가 아니면 불필요한 것 같아서 삭제하였습니다! 감사합니다 😊

yejineee replied:
불필요하여 삭제했습니닷...! 감사합니다 😊

Commit:
Fix: console.log 삭제

디버깅용으로 있던 console.log를 삭제한다.

Commit:
Style: sidebar border 및 dropdown item border 수정

사이드바의 외부 border는 디버깅 용이므로 삭제하고,
사이드바 컴포넌트 사이에 border-top을 주었다.
dropdown item에서도 자식 컴포넌트 사이에 border-top을 주었다.
위의 결과로 dropdown header와 dropdown 아이템 사이에 구분선이 필요하여,
dropdown header의 border-bottom을 추가하였다.

Commit:
Fix: sidebar index.js에서 api 호출 시  불필요한 클로저 함수 삭제

인자값을 받는 경우가 아니면, 클로저를 만들 이유가 없기 때문에,
바로 api 함수를 넘겨주는 것으로 수정하였다.

Commit:
Refactor: sidebar component 분리

사이드바에서 선택된 담당자, 라벨, 마일스톤을 렌더링하는 부분이
복잡하므로, 상위의 컨테이너로 분리시켰다.

Commit:
Style: dropdown profile 사이즈 변경

dropdown의 유저 프로필이 타원형이었다. 이를 수정하기 위해
유저 프로필의 크기를 줄이고, dropdown의 패딩 상하부분을 줄였다.

Commit:
Fix: dropdown 안에 있는 click handler에 id 전달

기존에는 dataset에 dropdown item의 id를 저장하였으나,
이는 html에 직접 드러나는 문제가 있어서, 클릭 핸들러의 인자로 id를 전달하는 것으로 수정한다.
